olive/passes/onnx/rtn_quantization.py

```diff
@@ -0,0 +1,349 @@
+        assert (block_size & (block_size - 1)) == 0, "Block size must be a power of 2."
+
+        quantize_axis = (quantize_axis + data_rank) % data_rank
+        quantized_data, scales, zero_points = self.quantize_ndarray(
+            data_ndarray, quantize_axis, block_size, self.is_symmetric
+        )
+
+        quantized_data_tensorproto = ir.Value(
+            name=node_initializer.name + f"_Q{self.bits}", const_value=ir.tensor(quantized_data)
+        )
+        scales_tensorproto = ir.Value(name=node_initializer.name + "_scales", const_value=ir.tensor(scales))
+        node_inputs = [quantized_data_tensorproto, node.inputs[1], scales_tensorproto]
+
+        if not self.is_symmetric:
+            zp_tensorproto = ir.Value(name=node_initializer.name + "_zero_points", const_value=ir.tensor(zero_points))
+            node_inputs.append(zp_tensorproto)
+
+        gather_axis = node.attributes.get_int("axis", 0)
+
+        kwargs = {
+            "gather_axis": gather_axis,
+            "quantize_axis": quantize_axis,
+            "block_size": block_size,
+        }
+
+        node.outputs[0].name = node.outputs[0].name + f"_Q{self.bits}"
+
+        return ir.node(
+            domain=MSFT_DOMAIN,
+            op_type=str(OpType.GatherBlockQuantized),
+            inputs=node_inputs,
+            name=node.name + f"_Q{self.bits}" if node.name else "",
+            attributes=kwargs,
+        )
+
+    def qbits_block_quant(self, fp32weight: npt.ArrayLike) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
+        """4b/8b quantize fp32 weight to int4 using C++ kernels."""
+        from onnxruntime.capi._pybind_state import quantize_matmul_4bits
+
+        qbits = self.bits
+        kpack = 8 // qbits
+        if len(fp32weight.shape) != 2:
+            raise ValueError("Current int4 block quantization only supports 2D tensors!")
+        rows, cols = fp32weight.shape
+
+        block_size = self.block_size
+        k_blocks = (rows + block_size - 1) // block_size
+
+        blob_size = (block_size + kpack - 1) // kpack
+        padded_rows = k_blocks * block_size
+        pad_len = padded_rows - rows
+        if pad_len > 0:
+            fp32weight = np.pad(fp32weight, ((0, pad_len), (0, 0)), "constant")
+
+        # block wise quantization, each block comes from a single column
+        packed = np.zeros((cols, k_blocks, blob_size), dtype="uint8")
+        zero_point = np.zeros(cols * ((k_blocks + kpack - 1) // kpack), dtype="uint8")
+        scales = np.zeros((cols * k_blocks), dtype=fp32weight.dtype)
+        if qbits == 8:
+            from onnxruntime import __version__ as OrtVersion
+            from packaging import version
+
+            if version.parse(OrtVersion) > version.parse("1.21.0"):
+                from onnxruntime.capi._pybind_state import quantize_matmul_8bits
+
+                quantize_matmul_8bits(packed, fp32weight, scales, zero_point, block_size, cols, rows, self.is_symmetric)
```

Comment:
I am slightly confused by this part. So this implementation is still in ort? I guess it needs to be packed then...